langkit/dsl_unparse.py

```diff
@@ -1,10 +1,177 @@
 from __future__ import absolute_import, division, print_function
 
 from langkit.passes import GlobalPass
+from contextlib import contextmanager
+
+try:
+    import libpythonlang as lpl
+except ImportError:
+    lpl = None
 
 templates = {}
 
 
+class DSLWalker(object):
+    if lpl is not None:
+        ctx = lpl.AnalysisContext()
+
+    class NoOpWalker(object):
+        def __init__(self):
+            pass
+
+        def __getattr__(self, item):
+            return self
+
+        def __call__(self, *args, **kwargs):
+            return self
+
+        def __enter__(self):
+            return self
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            return self
+
+        def emit_comments(self):
+            return ""
+
+    @staticmethod
+    def class_from_location(loc):
+        if lpl is None or loc is None:
+            return DSLWalker.NoOpWalker()
+
+        unit = DSLWalker.ctx.get_from_file(loc.file)
+        class_def = unit.root.find(
+            lambda n:
+            n.is_a(lpl.ClassDef) and
+            n.sloc_range.start.line == loc.line
+        )
+
+        if not class_def:
+            return DSLWalker.NoOpWalker()
+
+        return DSLWalker(class_def)
+
+    def __init__(self, from_node):
+        self.current_node = from_node
+        self.last_token = from_node.token_start
+        self.current_token = self.last_token
+
+    @contextmanager
+    def _with_current_node(self, new_node):
+        old_node = self.current_node
+        self.current_node = new_node
+        self.current_token = new_node.token_start
+        yield
+        self.current_node = old_node
+        self.current_token = new_node.token_end
+
+    def property(self, prop):
```

Comment:
Plz add a docstring for non annotated context managers saying that they're context managers